src/pages/batchImports/[resourceId]/index.js

```diff
@@ -0,0 +1,296 @@
+
+  if (lines.length === 0) return
+
+  const headers = lines[0].split(',').map(header => header.trim())
+
+  const columnMap = columns.reduce((map, col) => {
+    map[col.headerName] = col
+
+    return map
+  }, {})
+
+  const orderedColumns = headers.map(header => columnMap[header])
+
+  const rows = lines.slice(1).map(line => {
+    const values = line.split(',').map(value => value.trim())
+
+    return orderedColumns.reduce((obj, col, index) => {
+      const header = headers[index]
+      if (col) {
+        obj[col.field] = convertValue(values[index], col.dataType)
+      } else {
+        obj[header] = values[index]
+      }
+
+      return obj
+    }, {})
+  })
+
+  return transform(rows)
+}
+
+const getImportData = (gridData, columns, stackError) => {
+  const mandatoryColumns = columns.filter(col => col.mandatory)
+
+  const missingFields = gridData.list.flatMap(row =>
+    mandatoryColumns
+      .filter(col => row[col.field] === null || row[col.field] === undefined || row[col.field] === '')
+      .map(col => col.headerName)
+  )
+
+  if (missingFields.length > 0) {
+    const uniqueMissingFields = [...new Set(missingFields)]
+    stackError({
+      message: `${uniqueMissingFields.join(', ')} ${uniqueMissingFields.length > 1 ? 'are' : 'is'} mandatory field${
+        uniqueMissingFields.length > 1 ? 's' : ''
+      }.`
+    })
+
+    return
+  }
+
+  const convertedData = gridData.list.map(row => {
+    return Object.keys(row).reduce((acc, key) => {
+      const col = columns.find(c => c.field === key)
+      let value = row[key]
+      value = value === '' ? null : value
+      acc[key] = col ? convertValue(value, col.dataType, true) : value
+
+      return acc
+    }, {})
+  })
+
+  return convertedData
+}
+
+const BatchImports = () => {
+  const { stack } = useWindow()
+  const { stack: stackError } = useError()
+  const router = useRouter()
+
+  const [columns, setColumns] = useState([])
+  const [gridData, setGridData] = useState([])
+  const [name, setName] = useState('')
+  const [objectName, setObjectName] = useState('')
+  const [endPoint, setEndPoint] = useState('')
+
+  const { getRequest, postRequest } = useContext(RequestsContext)
+  const { resourceId } = router.query
+  const { platformLabels } = useContext(ControlContext)
+
+  const { access } = useResourceQuery({
+    datasetId: resourceId
+  })
+
+  useEffect(() => {
+    ;(async function () {
+      try {
+        if (resourceId) {
+          const res = await getRequest({
+            extension: SystemRepository.ETL.get,
+            parameters: `_resourceId=${resourceId}`
+          })
+
+          const modifiedFields = res.record.fields.map(({ name, dataType, ...rest }) => ({
+            field: name,
+            headerName: name,
+            type: (dataType == 2 || dataType == 3) && 'number',
+            flex: 1,
+            ...rest
+          }))
+
+          setColumns([{ field: 'recordId', headerName: '', flex: 0.4 }, ...modifiedFields])
```

Comment:
columns, objectName, and endpoint all derive from the response of the api. remove those states and have them be evaluated on every render (i.e. place the definition outside the useEffect).